hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/HazelcastClientCachingProvider.java

```diff
@@ -78,27 +78,31 @@ protected HazelcastInstance getOrCreateInstance(URI uri, ClassLoader classLoader
 
         // If config location is specified, get instance with its name.
         if (instanceName != null) {
-            return HazelcastClient.getHazelcastClientByName(instanceName);
+            HazelcastInstance instance = getOrCreateByInstanceName(instanceName);
+            return instance;
         }
 
         final boolean isDefaultURI = (uri == null || uri.equals(getDefaultURI()));
         if (!isDefaultURI) {
-            try {
-                // try locating a Hazelcast config at CacheManager URI
-                ClientConfig config = getConfigFromLocation(uri, classLoader, null);
-                return getOrCreateInstanceByConfig(config);
-            } catch (Exception e) {
-                if (LOGGER.isFinestEnabled()) {
-                    LOGGER.finest("Could not get or create hazelcast instance from URI " + uri.toString(), e);
+            if (isConfigLocation(uri)) {
```

Comment:
Also I see the server and client code is very similar. I am not sure if it (or parts of it) can be extracted in the superclass.